### PR TITLE
feat(appsync-modelgen-plugin): add null safety support in dart

### PR DESCRIPTION
**Pull request by AaronZyLee**
#### Description of changes

1. Add null safety support in dart modelgen
2. Add new feature flag `enableDartNullSafety`
3. Add a Dart sdk version validation function. The null safety will be disabled if the version is **less than 2.12.0**. Otherwise it depends on feature flag value.

#### Description of how you validated changes
`yarn test` and `amplify-dev codegen models`

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-codegen/blob/master/CONTRIBUTING.md#tests)
- [x] Relevant documentation is changed or added (and PR referenced)
- [x] Breaking changes to existing customers are released behind a feature flag or major version update
- [x] Changes are tested using sample applications for all relevant platforms (iOS/android/flutter/Javascript) that use the feature added/modified

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by phani-srikar — CHANGES_REQUESTED** [review]

File: packages/amplify-codegen/src/utils/validateDartSDK.js
Comment (phani-srikar) [inline packages/amplify-codegen/src/utils/validateDartSDK.js:8]: we should have unit test around this method

File: packages/amplify-codegen/src/utils/validateDartSDK.js
Comment (haverchuck) [inline packages/amplify-codegen/src/utils/validateDartSDK.js:11]: Does this semver package handle [caret syntax](https://dart.dev/tools/pub/dependencies#caret-syntax)?

File: packages/amplify-codegen/src/utils/validateDartSDK.js
Comment (haverchuck) [inline packages/amplify-codegen/src/utils/validateDartSDK.js:21]: So I believe that if the pubspec.yaml cannot be found or the version cannot be determined, we should make the codegen version determination using the feature flag.  I'm guessing that this warning should be changed to simply '\nCannot find dart sdk version' and the function can return true?

**Review by haverchuck — CHANGES_REQUESTED** [review]
Thanks for handling this! See inline comments.

File: packages/amplify-codegen/src/commands/models.js
Comment (phani-srikar) [inline packages/amplify-codegen/src/commands/models.js:77]: can we modify the FF to `enableDartNullSafety` as per https://github.com/aws-amplify/amplify-cli/pull/7516 just to be more clear that it is Dart specific.

File: packages/appsync-modelgen-plugin/src/__tests__/visitors/__snapshots__/appsync-dart-visitor.test.ts.snap
Comment (AaronZyLee) [reply packages/appsync-modelgen-plugin/src/__tests__/visitors/__snapshots__/appsync-dart-visitor.test.ts.snap:3376]: This is introduced and changed by a previous PR https://github.com/aws-amplify/amplify-codegen/pull/115

File: packages/appsync-modelgen-plugin/src/visitors/appsync-dart-visitor.ts
Comment (AaronZyLee) [reply packages/appsync-modelgen-plugin/src/visitors/appsync-dart-visitor.ts:261]: Yes they are optional in the field declaration no matter how they are defined except for `id` field, which is indicated by the proposed model

File: packages/appsync-modelgen-plugin/src/__tests__/visitors/__snapshots__/appsync-dart-visitor.test.ts.snap
Comment (AaronZyLee) [reply packages/appsync-modelgen-plugin/src/__tests__/visitors/__snapshots__/appsync-dart-visitor.test.ts.snap:3453]: Discussed offline. The element in the list can be `null` in some cases. The question mark will remain as before.

File: packages/amplify-codegen/src/utils/validateDartSDK.js
Comment (AaronZyLee) [reply packages/amplify-codegen/src/utils/validateDartSDK.js:11]: They do handle the caret version. https://www.npmjs.com/package/semver#ranges

File: packages/amplify-codegen/tests/commands/models.test.js
Comment (AaronZyLee) [inline packages/amplify-codegen/tests/commands/models.test.js:8]: The changes here are format-only conducted automatically by `prettier`. Please omit this for the review.

File: packages/amplify-codegen/tests/utils/validateDartSDK.test.js
Comment (haverchuck) [inline packages/amplify-codegen/tests/utils/validateDartSDK.test.js:32]: 👍

**Comment by haverchuck** [comment]
LGTM!  I'm not going to officially approve because I'm not familiar with all of this code, but the changes I've requested look good.

**Review by haverchuck — APPROVED** [review]
Approving, pending additional approval from codegen team.

File: packages/amplify-codegen/src/utils/validateDartSDK.js
Comment (Ashish-Nanda) [inline packages/amplify-codegen/src/utils/validateDartSDK.js]: Lets review the warning messages with the CLI and Library team.

Comment (fjnoyp) [reply packages/amplify-codegen/src/utils/validateDartSDK.js]: Shouldn't the warning of the message be "Detected" instead of "Detect"?  Detect might make it sound like you're telling the customer to do something.

**Comment by fjnoyp** [comment]
Hey the models are generated with 2020 instead of 2021

**Comment by fjnoyp** [comment]
I don't see any warnings about which Dart SDK is generated when I run the codegen locally.

**Comment by fjnoyp** [comment]
If I update a project's pubspec.yaml to Dart >= 2.12.0 from >= 2.7.0 I still get the original non null safe models?  Do you know why that might be happening?

**Comment by fjnoyp** [comment]
Hi @AaronZyLee in equals can we compare with the _ field instead of normal field.  The lines in red should be the desired codegen: 

image

**Comment by fjnoyp** [comment]
Also not required but if it's easy, can we update the toString method to use string interpolation like HuiSF described in his comment:

image

"id=$id, "

**Comment by AaronZyLee** [comment]
> Also not required but if it's easy, can we update the toString method to use string interpolation like HuiSF described in his comment:
> 
> image
> 
> "id=$id, "

There are cases where the toString value is not in the string form but an expression like list, enum. This is a more general way to generate the code otherwise will add more complexity.

**Review by fjnoyp — APPROVED** [review]
Looks good to me.  I'll give approval but let's please resolve the final warning messages for Dart SDK detection before merging this in.

File: packages/amplify-codegen/src/commands/models.js
Comment (cshfang) [inline packages/amplify-codegen/src/commands/models.js:88]: Sentence reads a little strangely. Maybe remove the "need to"?
```suggestion
        'Generating Dart Models without null safety. To generate null safe data models, turn on the “enabledartnullsafety” feature flag and set your Dart SDK version to “>= 2.12.0”.',
```

**Review by fjnoyp — APPROVED** [review]
Looks good.  Thanks for all your help in the back and forth changes we had for this.

**Review by phani-srikar — APPROVED** [review]

**Review by fjnoyp — CHANGES_REQUESTED** [review]
Hey Aaron sorry I found one last issue with the codegen models.  Now that we’ll be looking to truly load in nested models, there is a minor change needed in the fromJson method. 

Anytime we are creating a model field or list model field in fromJson, we need to call json[‘modelname’][‘serializedData’] instead of just json[‘modelname’]

Ex.
blog is a Blog field
comments is a [Comment] field

```
_blog = json['blog'] != null
    ? Blog.fromJson(
        new Map<String, dynamic>.from(json['blog']['serializedData']))
    : null,
```
Just add [‘serializedData’]

```
_comments = json['comments'] is List
    ? (json['comments'] as List)
        .map((e) => Comment.fromJson(new Map<String, dynamic>.from(e['serializedData'])))
        .toList()
    : null;
```
Just add [‘serializedData’]

**Review by fjnoyp — APPROVED** [review]

File: packages/amplify-codegen/src/commands/models.js
Comment (Ashish-Nanda) [inline packages/amplify-codegen/src/commands/models.js:86]: Do we need to set this to false?

File: packages/amplify-codegen/src/commands/models.js
Comment (haverchuck) [inline packages/amplify-codegen/src/commands/models.js:76]: Will this feature flag automatically get added to their cli.json prior to this code getting hit?

Comment (phani-srikar) [reply packages/amplify-codegen/src/commands/models.js:76]: Yes. If they are using the right CLI version (equal to above the version when this FF is added)

Comment (AaronZyLee) [reply packages/amplify-codegen/src/commands/models.js:76]: They will be added ahead as long as they are registered in CLI (aka Phani's PR gets merged and released)

File: packages/amplify-codegen/src/commands/models.js
Comment (AaronZyLee) [reply packages/amplify-codegen/src/commands/models.js:86]: If the minimum version is false and ff is true, it will parse true to modelgen unless we change it to false.

**Pull request merged**
